androidkeystore.cpp

```diff
@@ -0,0 +1,279 @@
+#include "androidkeystore_p.h"
+
+#include <QAndroidJniEnvironment>
+
+#include <QDebug>
+
+using namespace QKeychain;
+
+using namespace android::content;
+using namespace android::security;
+
+using namespace java::io;
+using namespace java::lang;
+using namespace java::math;
+using namespace java::util;
+using namespace java::security;
+using namespace java::security::spec;
+
+using namespace javax::crypto;
+using namespace javax::security::auth::x500;
+using namespace javax::security::cert;
+
+const BigInteger BigInteger::ONE = BigInteger::getStaticObjectField("java/math/BigInteger", "ONE", "Ljava/math/BigInteger;");
+
+const int Calendar::YEAR = Calendar::getStaticField<jint>("java/util/Calendar", "YEAR");
+
+const int Cipher::DECRYPT_MODE = Cipher::getStaticField<jint>("javax/crypto/Cipher", "DECRYPT_MODE");
+const int Cipher::ENCRYPT_MODE = Cipher::getStaticField<jint>("javax/crypto/Cipher", "ENCRYPT_MODE");
+
+namespace {
+
+QByteArray fromArray(const jbyteArray array)
+{
+    QAndroidJniEnvironment env;
+    jbyte *const bytes = env->GetByteArrayElements(array, Q_NULLPTR);
+    const QByteArray result(reinterpret_cast<const char *>(bytes), env->GetArrayLength(array));
+    env->ReleaseByteArrayElements(array, bytes, JNI_ABORT);
+    return result;
+}
+
+jbyteArray toArray(const QByteArray &bytes)
+{
+    QAndroidJniEnvironment env;
+    const int length = bytes.length();
+    jbyteArray array = env->NewByteArray(length);
+    env->SetByteArrayRegion(array, 0, length, reinterpret_cast<const jbyte *>(bytes.constData()));
+    return array;
```

Comment:
AFAIK NewByteArray holds a local reference to that object, so, IMHO it safer to return a QAndroidJniObject::fromLocalRef(array) instead